subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/verification/verifier/DependencyVerifierBuilder.java

```diff
@@ -222,8 +229,37 @@ public void addIgnoredKey(IgnoredKey key) {
             ignoredPgpKeys.add(key);
         }
 
-        public Set<String> buildTrustedPgpKeys() {
-            return pgpKeys;
+        /**
+         * Builds the list of trusted GPG keys.
+         * <p>
+         * This method will verify if all the trusted keys are in 160-bit fingerprint format.
+         * We do not accept either short or long formats, as they can be vulnerable to collision attacks.
+         *
+         * <p>
+         * Note: the fingerprints' formatting is not verified (i.e. if it's true base32 or not) at this stage.
+         * It will happen when these fingerprints will be converted to {@link org.gradle.security.internal.Fingerprint}.
+         *
+         * @return a set of trusted GPG keys
+         * @throws InvalidGpgKeyIdsException if keys not fitting the requirements were found
+         */
+        public Set<String> buildTrustedPgpKeys() throws InvalidGpgKeyIdsException {
+            final List<String> wrongPgpKeys = pgpKeys
+                .stream()
+                // The key is 160 bits long, encoded in base32 (case-insensitive characters).
+                //
+                // Base32 gives us 4 bits per character, so the whole fingerprint will be:
+                // (160 bits) / (4 bits / character) = 40 characters
+                //
+                // By getting ASCII bytes (aka. strictly 1 byte per character, no variable-length magic)
+                // we can safely check if the fingerprint is of the correct length.
+                .filter(key -> key.getBytes(StandardCharsets.US_ASCII).length != 40)
```

Comment:
Should we check for length `>=` just to be future proof?

Reply:
Good point, as e.g. [RFC 4880](https://www.rfc-editor.org/rfc/rfc4880#section-12.2) already states that this is V4 (i.e. it was changed 4 times already, with 2 notable changes).

`Fingerprint.java` only validates that the length of the fingerprint is divisible by 2 (you need 2 chars per byte), so we don't need to touch that.
https://github.com/gradle/gradle/blob/839bcd0137726151994ec8430ecfe8f0a0d9f61d/subprojects/security/src/main/java/org/gradle/security/internal/Fingerprint.java#L37-L40

So: yes, we can do that, though it breaks the error message:
> The following trusted GPG IDs are not in **160-bit** fingerprint format (see: ...)

Because in this case it's not guaranteed that it will be 160-bit, so I would suggest:
> The following trusted GPG IDs are not a fingerprint (see: ...)

Reply:
We probably can say "160-bit minimum" then

Reply:
Reworked this, so the error condition is  `length < 40`, change the exception message, and expanded the docs to explain the situation.